### DISCUSSION TITLE
why we need dotnet.wasm

### DISCUSSION THREAD

**Discussion by sajjadarashhh** [dotnet/runtime#94636] [Q&A] [answered]
why we need dotnet.wasm?
while browser that have owned runtime and we can execute wasm codes in that
if we compile c# modules directly to webassembly modules then why we need dotnet.wasm?

**Comment by hez2010** [comment, marked as answer]
Because currently we don't compile C# to webassembly modules directly. Instead we are using dotnet.wasm to interpret IL code.
You might find something interesting here: https://github.com/dotnet/runtimelab/tree/feature/NativeAOT-LLVM, which is the on-going work of compiling C# to webassembly directly with NativeAOT.

**Reply by Xyncgas** [reply]
That's why we need dotnet.wasm to run compiled .NET dll in webassembly running the donet rt

**Reply by sajjadarashhh** [reply]
When will we have this option in blazor wasm?

**Reply by LeaFrock** [reply]
Currently nobody is sure, maybe .NET 9/10 or not.

What does it matter to you? Do you meet perf bottleneck of dotnet.wasm?

**Reply by hez2010** [reply]
> Do you meet perf bottleneck of dotnet.wasm?

This one is obvious. dotnet.wasm is even 10x slower than JavaScript, while NativeAOT-LLVM can achieve native performance.

**Reply by Clockwork-Muse** [reply]
"But it's faster" doesn't actually mean "my code is currently too slow".

**Reply by sajjadarashhh** [reply]
@Clockwork-Muse in global benchmarking that says dotnet runtime is too slow in browser

**Reply by Clockwork-Muse** [reply]
@sajjadarashhh - 
That's not usually what benchmarks actually measure.  They normally give "this task takes this long/this many resources" answers, and then you have to decide if that makes sense for your usecase.

For instance, a month ago, I developed a small proof-of-concept that received events from an MQTT broker and displayed a bunch of stats on the screen.  I had a poor implementation due to the rush, but it worked well enough for the PoC.

**Reply by sajjadarashhh** [reply]
Size is very big 😐
We dont have critical issue in runtime we have many problem with startup time and first time load time
For example if you see mudblazor site you dont wana to open a pages in new tab for .net runtime load time... 😑🤦

**Reply by hez2010** [reply]
Slow speed + large size = no one is willing to use it in production (especially those websites which will be accessed by public users). 

- Slow speed: blazor costs a lot of time to initialize to load the runtime and render the first page, even if you have all those wasm files cached. This is very easy to observe on a low-performance endpoint such as smart phones.
- Large size: spend more money for delivering contents from a CDN because it requires more network traffic.

The WebSocket based interactive server-side rendering doesn't resolve anything as it only moves the workload from clients to servers, which will cost more compute resources, and it also makes it impossible to scale out/restart the server seamlessly because all states will be lost if the server down. 

Without NativeAOT-LLVM, blazor is only suitable to develop projects for internal use (i.e. those projects which are not intended to be accessed by public users) for both performance and size reason.
